Post a Playground preview link on every PR

adamziel opened the pull request:
Very simple GitHub workflow to comment on every `trunk` PR with a link to a Playground-based preview.

Other than access, it seems to be fine. The last thing to solve is the 403 error:

> 403 Resource not accessible by integration

https://github.com/WordPress/wordpress-develop/actions/runs/6572036562/job/17852349020?pr=5526

cc @jeffpaul 

Trac Ticket: https://core.trac.wordpress.org/ticket/59416.

adamziel committed:
Post a Playground preview link on every PR

adamziel committed:
Adjust workflow name

adamziel committed:
Run on all pull requests

adamziel committed:
Try a different approach

adamziel committed:
Remove "identifier" variable

jeffpaul commented:
This action adds a comment, so perhaps there's something to glean from that to leverage here?  https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/welcome-new-contributors.yml

cc: @desrosj

adamziel committed:
Update preview-comment.yml

adamziel committed:
Update preview-comment.yml

desrosj commented on .github/workflows/preview-comment.yml:
```diff
@@ -0,0 +1,31 @@
+    steps:
+      - uses: actions/github-script@v6
+        env:
+          GITHUB_TOKEN: ${{ env.BOT_TOKEN }}
```
We shouldn't need to supply a token. The default `$GITHUB_TOKEN` should be sufficient.

swissspidy replied:
Not with the `pull_request` trigger. The token is read-only. With `pull_request_target` it should work.

Alternatively a custom fine-grained PAT would need to be provided.

adamziel replied:
I updated it to use the `pull_request_target`, thanks!

desrosj commented on .github/workflows/preview-comment.yml:
```diff
@@ -0,0 +1,31 @@
+        with:
+          github-token: ${{ env.BOT_TOKEN }}
+          script: |
+            const previewLink = `https://playground.wordpress.net/wordpress.html?pr=${context.issue.number}`;
```
Can we add a bit more context to this? If I am someone opening a PR for the first time and have no idea what this links to or what Playground even is, I should have a brief explanation in the comment and links out to some resources to learn more.

adamziel replied:
Absolutely! It’s only brief because I focused on auto-posting which doesn’t work yet. Let’s flesh out this PR a bit more, then

adamziel replied:
How does the updated version work for you @desrosj ?

desrosj replied:
I've made some additional adjustments. Please have a look and see!

desrosj reviewed (changes requested):
I don't think we should have a separate workflow for this. I think that this should be a step within the Build WordPress workflow after the artifact is built and uploaded.

swissspidy commented on .github/workflows/preview-comment.yml:
```diff
@@ -0,0 +1,31 @@
+name: Comment on PR
+
+on:
+  pull_request:
+    branches:
+      - trunk
```
This would cause the comment to be posted multiple times, whenever a PR is updated.

I'd change it something like this:

```
on:
  pull_request_target:
    types: [ opened ]
    branches:
      - trunk
```

adamziel committed:
Document the workflow better and change the trigger to pull_request_target

adamziel committed:
Merge branch 'trunk' into preview-comment-on-every-pr

adamziel committed:
Remove target: [ insert ] for a second to actually post a comment

adamziel committed:
Added types: [edited] to add the comment to the current PR

adamziel committed:
Add a few more acticity types to test the PR

adamziel committed:
Set the action types to just opened

adamziel commented:
Thanks you @desrosj @swissspidy! I just addressed the feedback on this PR, although we won't able to test it without merging. [GitHub actions docs says](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target):

> This event runs in the context of the base of the pull request, rather than in the context of the merge commit, as the pull_request event does. This prevents execution of unsafe code from the head of the pull request that could alter your repository or steal any secrets you use in your workflow. This event allows your workflow to do things like label or comment on pull requests from forks. Avoid using this event if you need to build or run code from the pull request.

adamziel commented:
> I don't think we should have a separate workflow for this. I think that this should be a step within the Build WordPress workflow after the artifact is built and uploaded.

This workflow now uses a different trigger (`pull_request_target`) than the Build WordPress workflow so it may need to remain separate.

swissspidy commented:
@adamziel you can test this by opening a PR on your own fork instead this repository. This way we can refine it before merging.

We can also try using a different trigger with a custom access token if we want to merge the workflow files.

desrosj commented:
What was the reason pull request wouldn't work? I think I'm missing that.

swissspidy commented:
@desrosj correct me if I‘m wrong, but permissions are different for that trigger. See https://github.com/WordPress/wordpress-develop/pull/5526#discussion_r1370397098

desrosj commented:
Ah, sorry I missed that because it was collapsed in the resolved discussion. We should be able to use `permissions` to fine tune the capability we need the workflow, job, or step to have. https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs#overview

desrosj committed:
Try using `pull_request` with fine-tuned permissions.

desrosj committed:
Remove `opened` type filter to test commenting.

desrosj committed:
Use `issues` permission for fine tuning.

desrosj committed:
Add `write-all` to see if using `permissions` will even work.

desrosj committed:
Change back to `pull_request_target`.

desrosj committed:
Remove `opened` type filter to test.

desrosj committed:
Combine two related workflows into one and rename to WordPress Playground

desrosj committed:
Correct permissions for comment job.

desrosj committed:
Improve comment wording.

desrosj committed:
Add `pull_request` just to test.

desrosj committed:
Attempt to resolve invalid workflow warning.

desrosj committed:
Attempt to resolve invalid workflow warning, take 2.

desrosj committed:
Add a link to the limitations page in the docs.

desrosj committed:
Remove `pull_request` event that was just for testing.

swissspidy commented on .github/workflows/wordpress-playground.yml:
```diff
@@ -0,0 +1,106 @@
+
+            For more details about these limitations and more, check out the [Limitations page](https://wordpress.github.io/wordpress-playground/limitations/) in the WordPress Playground documentation.
+
+            [Test this pull request WordPress Playground](https://playground.wordpress.net/wordpress.html?pr=${context.issue.number}).
```
```suggestion
            [Test this pull request with WordPress Playground](https://playground.wordpress.net/wordpress.html?pr=${context.issue.number}).
```

desrosj committed:
Detect the presence of a comment.

desrosj committed:
Fix typo.

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

desrosj committed:
Add conditions to the commenting job to only run on wordpress-develop.

desrosj committed:
Update conditions using `pull_request` to `pull_request_target` instead.

desrosj commented:
Sorry for all the noise!

After some testing, I think you're right @swissspidy. I could have sworn that in the past I was able to get commenting to work within a PR, but I must be mixing that up either with workflows that are _triggered_ by other workflows within a PR (and thus run within the context of the repo itself) or instances where I had a PR from a branch within the same repository and not a fork.

I still feel like we should not have a separate workflow just for commenting. The commenting is related to the build step, and even depends on it (we shouldn't comment unless we know that a build has successfully been created). I've gone and combined the two, and also renamed `build.yml` to `wordpress.playground.yml`, which will allow us to include other additional logic related to Playground in the future.

A few other notes about the changes I included:
- I pinned a [specific commit SHA to the version](https://github.com/WordPress/wordpress-develop/pull/5526/files#diff-63e16d2e0afb024b814b40c6e31d664ace2208a6d2656e944c700c5a2f3ceba8R80) of actions/github-scripts. This is a recommended best practice in [GitHub's hardening recommendations when using third-party actions](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions).
- I've made the commenting job dependent on the build one. We shouldn't comment on the PR until there actually is an artifact that can be used. This logic is untested though since we can't verify until merged. So please review extra carefully.
- I removed the `opened` type filter for the event as well. This ensures that the build runs on every update to the PR.
- To prevent comments being posted every time the workflow runs, I've changed the commenting job to look for a preexisting comment before leaving a new one. This also solves the edge case where the first build attempt for a PR fails. The comment job will skip the first time, and it will run once there is a successful build.
- I expanded the comment a bit to call out some of the limitations and "gotchas" about using a Playground instance. I also linked off to the Limitations page in the documentation so someone can dig deeper if they would like.

One other thought. I'm not sure that running on `workflow_dispatch` is useful if a Playground instance cannot be created using a commit hash. Something like `https://playground.wordpress.net/wordpress.html?sha=abcdefghijklmnop`. Even if that's possible, I think that adding a `push` trigger event would be better. That would ensure every commit to `trunk` (or whatever branches we specify) has a built artifact that could be used in a Playground instance.

desrosj commented:
Also, here is an example of what the content of the comment will look like:

## Test using WordPress Playground
The changes in this pull request can previewed and tested using a [WordPress Playground](https://developer.wordpress.org/playground/) instance.

[WordPress Playground](https://developer.wordpress.org/playground/) is an experimental project that creates a full WordPress instance entirely within the browser.

### Some things to be aware of
- The Plugin and Theme Directories cannot be accessed within Playground.
- All changes will be lost when closing a tab with a Playground instance.
- All changes will be lost when refreshing the page.
- A fresh instance is created each time the link below is clicked.
- Every time this pull request is updated, a new ZIP file containing all changes is created. If changes are not reflected in the Playground instance, it's possible that the most recent build failed, or has not completed. Check the [list of workflow runs to be sure](https://github.com/WordPress/wordpress-develop/actions/workflows/wordpress-playground.yml).

For more details about these limitations and more, check out the [Limitations page](https://wordpress.github.io/wordpress-playground/limitations/) in the WordPress Playground documentation.

[Test this pull request WordPress Playground](https://playground.wordpress.net/wordpress.html?pr=5526).

desrosj committed:
Minor typo fixes.

swissspidy commented:
> I've made the commenting job dependent on the build one. We shouldn't comment on the PR until there actually is an artifact that can be used. This logic is untested though since we can't verify until merged.

And why can't we test this by opening a PR against the fork? That's how I successfully tested this in the past. This comment is created by the GitHub Action: https://github.com/swissspidy/wordpress-develop/pull/33

desrosj committed:
Merge branch 'preview-comment-on-every-pr' of github.com:adamziel/wordpress-develop into preview-comment-on-every-pr

desrosj committed:
Remove conditions preventing runs.

desrosj committed:
Temporary branch for running.

adamziel commented:
I'm on a meetup this week and AFK the next one – feel absolutely free to take over here

desrosj committed:
Correct syntax for setting the comment body.

desrosj committed:
Merge remote-tracking branch 'upstream/trunk' into preview-comment-on-every-pr

desrosj committed:
ZIP the built directory and upload as an artifact.

desrosj committed:
Remove path filtering for pull requests.

Because this is required for building a zip artifact, it must run on all pull requests.

desrosj committed:
Merge Welcome and Playground comment workflows.

desrosj committed:
Add workflow call from test workflow and cleanup.

adamziel commented:
@desrosj did you manage to get it to work in your fork?

desrosj commented:
I did, but I have a bit of cleanup to do. I plan to circle back to this later this week after 6.4 is out!

desrosj committed:
Merge improvements.

desrosj committed:
Revert a few test changes.

desrosj commented:
OK, I think I've cleaned up my work and have it ready to go. You can see a test comment in desrosj/wordpress-develop#141.

I chatted with @swissspidy, and thought about trying this way forward. We are already building WordPress in the workflow that tests the build process. So it makes sense to just perform the zipping and uploading of the artifact in that workflow instead of having a completely different one. That doesn't solve the issue with requiring `pull_request_target` though.

To solve this, I've consolidated the welcome comment workflow with the logic that comments with Playground details. It will run when a PR is opened on `pull_request_target` for the welcome message, but it will only post the comment with the Playground details after the build workflow has succeeded. This ensures that there is an artifact present that can actually be used. This is triggered using the `workflow_dispatch` event, which does not have the same issue as the `pull_request` event.

I don't love that it will run every time a PR is updated, but I'm not sure any way around that.

adamziel commented:
@desrosj I just reviewed and this looks good to me, thank you so much for taking the lead here! It indeed a pity this needs to run on every PR update, but I don't think it's a big deal. I like how you consolidated that with the new contributor comment 🎉  Let's ship it!

swissspidy reviewed (approved):
Let's give it a go!

desrosj commented:
Merged in https://core.trac.wordpress.org/changeset/57124. However, it seems to be having issues dispatching the workflow: https://github.com/WordPress/wordpress-develop/actions/runs/6907809373/job/18797346079#step:2:82.

adamziel commented:
@desrosj weird! I see the proper `permissions` are in place inside the workflow. [Someone on StackOverflow](https://stackoverflow.com/questions/70435286/resource-not-accessible-by-integration-on-github-post-repos-owner-repo-ac) suggested this:

> go to https://github.com/OWNER/REPO/settings/actions and in Workflow Permissions section give actions Read and Write permissions. That provides your token with rights to modify your repo and solves your problem.

adamziel commented:
@desrosj interestingly, the "first contributor" comment still works. Also, dispatching the "Pull Request Comments" workflow manually seems to have worked – I attribute that to the `workflow_dispatch` condition but I'll try a few more things.

See my test PR: https://github.com/WordPress/wordpress-develop/pull/5685.

adamziel commented:
So [workflow_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) doesn't come with the same note about permissions as  [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target):

> For workflows that are triggered by the pull_request_target event, the GITHUB_TOKEN is granted read/write repository permission

What if we just lean on `pull_request_target` and add a message on WordPress Playground side like "this Pull Request is still being built – please wait a few moments"?

adamziel commented:
I found one more issue with the uploaded artifact – it's 332 bytes and doesn't actually contain the built files:

<img width="647" alt="CleanShot 2023-11-25 at 11 46 32@2x" src="https://github.com/WordPress/wordpress-develop/assets/205419/d90e60f8-40c9-4229-b3bb-b7e6855a0243">

adamziel commented:
This was merged in https://github.com/WordPress/wordpress-develop/pull/5737
